### PR TITLE
Decomps

### PR DESCRIPTION
**Pull request by alewis**
- foo_decomposition now called foo
- split_axis -> pivot_axis
- qr and rq accept a non_negative_diagonal option that enforces R have this property

### CODE REVIEW

File: tensornetwork/backends/tensorflow/tensorflow_backend.py [hunk 1]
```diff
@@ -58,29 +58,29 @@ def slice(self, tensor: Tensor, start_indices: Tuple[int, ...],
                        "identical.")
     return tf.slice(tensor, start_indices, slice_sizes)
 
-  def svd_decomposition(
+  def svd(
       self,
       tensor: Tensor,
-      split_axis: int,
+      pivot_axis: int = 1,
```
Comment (Thenerdstation) [inline tensornetwork/backends/tensorflow/tensorflow_backend.py:64]: Is this consistent default? I assumed it would split on the last axis.

Comment (alewis) [reply tensornetwork/backends/tensorflow/tensorflow_backend.py:64]: It's so that pivot is a null-op on matrices: 
A = np.ones((4, 2))
B = A.reshape( (A.shape[:1], A.shape[1:]. ) ) # new shape is also (4, 2)

Comment (alewis) [reply tensornetwork/backends/tensorflow/tensorflow_backend.py:64]: -1 would also behave this way on matrices, if you prefer that

Comment (Thenerdstation) [reply tensornetwork/backends/tensorflow/tensorflow_backend.py:64]: I think i'm slightly more in favor of -1.

File: tensornetwork/backends/pytorch/decompositions.py [hunk 2]
```diff
@@ -146,38 +147,43 @@ def qr_decomposition(
...
-  right_dims = list(tensor.shape)[split_axis:]
+  left_dims = list(tensor.shape)[:pivot_axis]
+  right_dims = list(tensor.shape)[pivot_axis:]
 
   tensor = torch.reshape(tensor, (np.prod(left_dims), np.prod(right_dims)))
   q, r = torch.qr(tensor)
+  if non_negative_diagonal:
+    phases = torch.sign(torch.diagonal(r))
+    q = q * phases
+    r = torch.diag_embed(phases) @ r
```
Comment (mganahl) [inline tensornetwork/backends/pytorch/decompositions.py:166]: I think you can use broadcasting here as well:
`phases[:,None] * r` should be the same (check again though)

File: tensornetwork/backends/pytorch/decompositions.py [hunk 3]
```diff
@@ -190,19 +196,23 @@ def rq_decomposition(
...
-  right_dims = tensor.shape[split_axis:]
+  left_dims = tensor.shape[:pivot_axis]
+  right_dims = tensor.shape[pivot_axis:]
   tensor = torch.reshape(tensor, [np.prod(left_dims), np.prod(right_dims)])
   #torch has currently no support for complex dtypes
   q, r = torch.qr(torch.transpose(tensor, 0, 1))
+  if non_negative_diagonal:
+    phases = torch.sign(torch.diagonal(r))
+    q = q * phases
+    r = torch.diag_embed(phases) @ r
```
Comment (mganahl) [inline tensornetwork/backends/pytorch/decompositions.py:215]: use broadcasting

File: tensornetwork/backends/symmetric/decompositions.py [hunk 4]
```diff
@@ -193,14 +193,13 @@ def svd_decomposition(
       discarded_singvals > 0.0]
 
 
-def qr_decomposition(bt, tensor: BlockSparseTensor,
-                     split_axis: int) -> Tuple[Tensor, Tensor]:
+def qr(bt, tensor: BlockSparseTensor, pivot_axis: int) -> Tuple[Tensor, Tensor]:
```
Comment (mganahl) [inline tensornetwork/backends/symmetric/decompositions.py:196]: can you add the non_negative_diagonal here and raise an error if its True?

File: tensornetwork/backends/symmetric/decompositions.py [hunk 5]
```diff
@@ -209,14 +208,13 @@ def qr_decomposition(bt, tensor: BlockSparseTensor,
   return q, r
 
 
-def rq_decomposition(bt, tensor: BlockSparseTensor,
-                     split_axis: int) -> Tuple[Tensor, Tensor]:
+def rq(bt, tensor: BlockSparseTensor, pivot_axis: int) -> Tuple[Tensor, Tensor]:
```
Comment (mganahl) [inline tensornetwork/backends/symmetric/decompositions.py:211]: same here

Comment (mganahl) [reply tensornetwork/backends/symmetric/decompositions.py:211]: nevermind

File: tensornetwork/backends/tensorflow/decompositions.py [hunk 6]
```diff
@@ -151,39 +152,44 @@ def qr_decomposition(
...
+  right_dims = tf.shape(tensor)[pivot_axis:]
 
   tensor = tf.reshape(tensor,
                       [tf.reduce_prod(left_dims),
                        tf.reduce_prod(right_dims)])
   q, r = tf.linalg.qr(tensor)
+  if non_negative_diagonal:
+    phases = tf.math.sign(tf.linalg.diag_part(r))
+    q = q * phases
+    r = tf.linalg.diag(phases) @ r
```
Comment (mganahl) [inline tensornetwork/backends/tensorflow/decompositions.py:172]: use broadcasting

File: tensornetwork/backends/tensorflow/decompositions.py [hunk 7]
```diff
@@ -196,20 +202,24 @@ def rq_decomposition(
...
+  right_dims = tf.shape(tensor)[pivot_axis:]
 
   tensor = tf.reshape(tensor,
                       [tf.reduce_prod(left_dims),
                        tf.reduce_prod(right_dims)])
   q, r = tf.linalg.qr(tf.math.conj(tf.transpose(tensor)))
+  if non_negative_diagonal:
+    phases = tf.math.sign(tf.linalg.diag_part(r))
+    q = q * phases
+    r = tf.linalg.diag(phases) @ r
```
Comment (mganahl) [inline tensornetwork/backends/tensorflow/decompositions.py:222]: broadcasting

File: tensornetwork/backends/tensorflow/tensorflow_backend.py (hunk 1 above)
Comment (alewis) [reply tensornetwork/backends/tensorflow/tensorflow_backend.py]: Okay, I changed it to -1. I think that's a bit better too on reflection

File: tensornetwork/backends/pytorch/decompositions.py (hunk 2 above)
Comment (alewis) [reply tensornetwork/backends/pytorch/decompositions.py]: Works

File: tensornetwork/backends/symmetric/decompositions.py (hunk 4 above)
Comment (alewis) [reply tensornetwork/backends/symmetric/decompositions.py:196]: This happens at the backend level

**Comment by alewis** [comment]
Comments should be addressed

**Comment by mganahl** [comment]
fix conflicts and we can merge

**Comment by alewis** [comment]
Fixed

File: tensornetwork/backends/pytorch/decompositions.py [hunk 8]
```diff
@@ -146,38 +147,44 @@ def qr_decomposition(
...
+  left_dims = list(tensor.shape)[:pivot_axis]
+  right_dims = list(tensor.shape)[pivot_axis:]
 
   tensor = torch.reshape(tensor, (np.prod(left_dims), np.prod(right_dims)))
   q, r = torch.qr(tensor)
+  if non_negative_diagonal:
+    phases = torch.sign(torch.diagonal(r))
+    q = q * phases
+    r = phases[:, None] * r
+    #r = torch.diag_embed(phases) @ r
```
Comment (mganahl) [inline tensornetwork/backends/pytorch/decompositions.py:167]: remove commented code

Comment (alewis) [reply tensornetwork/backends/pytorch/decompositions.py]: done
